### PR TITLE
 Kubernetes deployment initial version

### PR DESCRIPTION
**Pull request by campidelli**

### CODE REVIEW

File: docker/k8s_build_push_run.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+printf "\n\n******** Bulding Java Projects of GroupID ********\n\n"
+pwd
+cd ..
+#mvn clean install
```
Comment (robson-sanchez) [inline docker/k8s_build_push_run.sh]: Shouldn't it be uncommented?

File: helpdesk-bot/pom.xml [hunk 2]
```diff
@@ -103,11 +123,39 @@
             <version>1.58</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.bouncycastle</groupId>
```
Comment (robson-sanchez) [inline helpdesk-bot/pom.xml]: The same dependency declared twice conflicting versions.

File: helpdesk-bot/docker/k8s_deployment.yaml.template [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            - name: HELPDESK_SERVICE_HOST
+              value: helpdesk-api-service
+            - name: HELPDESK_SERVICE_PORT
+              value: '80'
+            - name: AGENT_HOST
+              value: nexus2-2.symphony.com
+            - name: AGENT_PORT
+              value: '443'
+            - name: SESSION_AUTH_HOST
+              value: sym-nexus2-dev-chat-glb-3-ause1-all.symphony.com
```
Comment (robson-sanchez) [inline helpdesk-bot/docker/k8s_deployment.yaml.template:38]: All the service endpoints, username, keystores, passwords, and streamId are hard-coded. Shouldn't it be configured externally via environment variables or something like that?

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrap.java [hunk 4]
```diff
@@ -73,6 +75,7 @@ public void execute(ApplicationContext applicationContext) {
         .onError(e -> LOGGER.error("Fail to initilize Helpdesk Ai", e));
 
     try {
+      executeProvisioning(applicationContext);
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrap.java]: You should execute this using FunctionExecutor to ensure retry policy will be applied.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+/**
+ * Provides acess to public APIs of a POD.
+ * Created by campidelli on 03/12/17.
+ */
+@Service
+public class HelpDeskPublicApiClient {
+
+  @Autowired
+  HelpDeskBotConfig config;
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: You might define this field as private and move it below static fields.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+   * @return Generated salt.
+   */
+  public String getSalt(String userName) {
+    String endpoint = config.getPod().getUrl(GET_SALT_ENDPOINT);
+    endpoint = String.format(endpoint, userName);
+
+    ResponseEntity<String> response = restTemplate.getForEntity(endpoint, String.class);
+
+    String json = response.getBody();
+    ObjectMapper mapper = new ObjectMapper();
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: The ObjectMapper object is too expensive to be created every time you call this method. Please move it as constant field and perform an eager initialization.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+   * @return Session key.
+   */
+  public String login(String userName, String encryptedPassword) {
+    String endpoint = config.getPod().getUrl(LOGIN_ENDPOINT);
+
+    MultiValueMap<String, Object> formData = new LinkedMultiValueMap<>();
+    formData.add("userName", userName);
+    formData.add("hPassword", encryptedPassword);
+
+    ResponseEntity<String> response = restTemplate.postForEntity(endpoint, formData, String.class);
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: You must verify HTTP status code to deal with HTTP errors.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    List<String> cookies = response.getHeaders().get(HttpHeaders.SET_COOKIE);
+    String cookie = cookies.stream().filter(c -> c.startsWith(SESSION_KEY)).findFirst().get();
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: Please verify if the cookie already exists before calling the 'get' method.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    headers.setContentType(MediaType.APPLICATION_JSON);
+
+    HttpEntity<CompanyCert> entity = new HttpEntity<CompanyCert>(certificate, headers);
+
+    String endpoint = config.getPod().getUrl(CREATE_CERTIFICATE_ENDPOINT);
+
+    ResponseEntity<CompanyCertDetail> response = restTemplate.postForEntity(
+        endpoint, entity, CompanyCertDetail.class);
+
+    return response.getBody();
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java:117]: You must verify HTTP status code to deal with HTTP errors.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * Created by campidelli on 3/9/18.
+ */
+@Service
+public class HelpDeskProvisioningService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(HelpDeskProvisioningService.class);
+
+  @Autowired
+  ProvisioningConfig config;
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: You might define all fields as private fields.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+  @Autowired
+  HelpDeskPublicApiClient publicApiClient;
+
+  @Autowired
+  CertificateUtils certificateUtils;
+
+  public void execute() {
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: Javadoc.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    if (!config.isExecute()) {
+      LOGGER.info("Provisioning process will not be executed.");
+      return;
+    }
+    LOGGER.info("Provisioning is being executed.");
+
+    // Root certificate
+    if (config.isGenerateCACert()) {
+      LOGGER.info("Generating a Self signed CA ROOT certificate.");
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: Please extract this behavior to a method.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+        String sessionToken = login();
+        LOGGER.info("Logged in successfully. Uploading certificate.");
+        publicApiClient.createCompanyCert(sessionToken, cert);
+        LOGGER.info("Certificate successfully uploaded.");
+      }
+    }
+
+    // Service account
+    if (config.isGenerateServiceAccountP12()) {
+      LOGGER.info("Generating p12 file for a service account.");
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: Please extract this behavior to a method.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+      certificateUtils.createUserCertificate(caKeyPath, caCertPath, serviceAccountUserName);
+      LOGGER.info("p12 file successfully generated.");
+    }
+  }
+
+  private CompanyCert getSelfSignedCertificate() {
+    X509Certificate certificate = certificateUtils.createSelfSignedRootCertificate();
+    String pem = certificateUtils.getPemAsString(certificate);
+    String fileName = certificateUtils.getSelfSignedRootCertificatePath();
+    File file = new File(fileName);
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java:86]: Is this file required?

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/TestContext.java [hunk 15]
```diff
@@ -1,4 +1,4 @@
-package org.symphonyoss.symphony.bots.helpdesk.bot.it;
+package org.symphonyoss.symphony.bots.helpdesk.bot.util;
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/TestContext.java]: It should be part of test source code or renamed.

File: helpdesk-bot/src/main/resources/application.yaml [hunk 16]
```diff
@@ -48,3 +48,13 @@ app-authentication:
 http-client:
   connect-timeout: 45000
   read-timeout: 45000
+
+provisioning:
+  execute: false
+  user:
+    name: <user_with_provisioning_role>
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/resources/application.yaml]: You should define here only default values. Settings that depends on the environment shouldn't be part of the application classpath.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrapTest.java [hunk 17]
```diff
@@ -89,6 +93,7 @@ public void init() {
     doReturn(helpDeskBot).when(applicationContext).getBean(HelpDeskBot.class);
     doReturn(helpDeskAi).when(applicationContext).getBean(HelpDeskAi.class);
     doReturn(httpClient).when(applicationContext).getBean(HelpDeskHttpClient.class);
+    doReturn(provisioningService).when(applicationContext).getBean(HelpDeskProvisioningService.class);
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrapTest.java:112]: You didn't test this new stuff.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+import java.io.IOException;
+import java.net.HttpCookie;
+import java.util.List;
+
+/**
+ * Provides acess to public APIs of a POD.
+ * Created by campidelli on 03/12/17.
+ */
+@Service
+public class HelpDeskPublicApiClient {
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java:27]: You missed unit tests for this class.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/config/HelpDeskBotConfig.java [hunk 19]
```diff
@@ -235,7 +236,10 @@ public void setGroupId(String groupId) {
   }
 
   public String getAgentStreamId() {
-    return agentStreamId;
+    if (agentStreamId != null) {
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/config/HelpDeskBotConfig.java:239]: You missed testing this method.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+
+/**
+ * Performs the provisioning process of a bot. It consists of:
+ *
+ * - Generate a self-signed certificate and submit it to a POD;
+ *
+ * Created by campidelli on 3/9/18.
+ */
+@Service
+public class HelpDeskProvisioningService {
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: You missed unit tests for this class.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/CertificateUtils.java [hunk 21]
```diff
@@ -0,0 +1,449 @@
...
+
+import javax.annotation.PostConstruct;
+
+/**
+ * Utility class to deal with certificate stuff.
+ *
+ * Created by crepache on 01/03/18.
+ */
+@Component
+public class CertificateUtils {
```
Comment (robson-sanchez) [inline helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/CertificateUtils.java:82]: You missed unit tests for this class.

**Review by robson-sanchez — CHANGES_REQUESTED** [review]

File: docker/k8s_build_push_run.sh (hunk 1 above)
Comment (campidelli) [reply docker/k8s_build_push_run.sh]: Yes, it should!

File: helpdesk-bot/pom.xml (hunk 2 above)
Comment (campidelli) [reply helpdesk-bot/pom.xml]: Done.

File: helpdesk-bot/docker/k8s_deployment.yaml.template (hunk 3 above)
Comment (campidelli) [reply helpdesk-bot/docker/k8s_deployment.yaml.template:38]: This file is not part of the Dcoker image. It is used as a deployment descriptor on kubernetes. There must be one of this file for each environment. By the way, these pair of name/value become environment variables when the container is running.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrap.java (hunk 4 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrap.java]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java (hunk 5 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrapTest.java (hunk 17 above)
Comment (campidelli) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/bootstrap/HelpDeskBootstrapTest.java:112]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java (hunk 18 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java:27]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/config/HelpDeskBotConfig.java (hunk 19 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/config/HelpDeskBotConfig.java:239]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java (hunk 20 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/CertificateUtils.java (hunk 21 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/CertificateUtils.java:82]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java (hunk 7 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: It throws a RuntimeException when code is different from 2XX.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java (hunk 6 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java (hunk 8 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java (hunk 9 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/client/HelpDeskPublicApiClient.java:117]: It throws a RuntimeException when code is different from 2XX.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java (hunk 12 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java (hunk 13 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java (hunk 14 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java:86]: Yes because the ```String fileName``` holds the whole path and I need the file name only.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/TestContext.java (hunk 15 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/util/TestContext.java]: Renamed!

File: helpdesk-bot/src/main/resources/application.yaml (hunk 16 above)
Comment (campidelli) [reply helpdesk-bot/src/main/resources/application.yaml]: Done.

File: helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java (hunk 10 above)
Comment (campidelli) [reply helpdesk-bot/src/main/java/org/symphonyoss/symphony/bots/helpdesk/bot/provisioning/HelpDeskProvisioningService.java]: Done.

**Review by robson-sanchez — APPROVED** [review]
